Assets/Scripts/GameStateMachine/GameStateMachine.cs

```diff
@@ -0,0 +1,80 @@
+using System.Collections.Generic;
+using System.Linq;
+
+public static class GameStateMachine
+{
+    public static GameStateMachineBehaviour OwningGameStateMachine { get; private set; }
+    public static List<State> States { get; } = new List<State>();
+    public static State CurrentState { get; private set; }
+
+    /// <summary>
+    /// Allows States to subscribe themselves to the StateMachine.
+    /// </summary>
+    /// <param name="state">The State subscribing itself to the StateMachine.</param>
+    public static void Subscribe(this State state)
+    {
+        for (int i = 0; i < States.Count; i++)
+            if (States[i].ToString() == state.ToString())
+                return;
```

Comment:
```suggestion
        for (int i = 0; i < States.Count; i++)
        {
            if (States[i].ToString() == state.ToString())
                return;
        }
```
I like it when something without a scope that does technically has another scope in it to add a scope then.

Reply:
This better?
```csharp
    public static void Subscribe(this State state)
    {
        if (!States.Any(s => s.ToString() == state.ToString()))
            States.Add(state);
    }
   ```